datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -3633,32 +3697,123 @@ mod tests {
 
     #[test]
     fn test_like_and_ilke() {
-        // LIKE '%'
-        let expr = like(col("c1"), "%");
+        let null = lit(ScalarValue::Utf8(None));
+
+        // expr [NOT] [I]LIKE NULL
+        let expr = like(col("c1"), null.clone());
+        assert_eq!(simplify(expr), lit_bool_null());
+
+        let expr = not_like(col("c1"), null.clone());
+        assert_eq!(simplify(expr), lit_bool_null());
+
+        let expr = ilike(col("c1"), null.clone());
+        assert_eq!(simplify(expr), lit_bool_null());
+
+        let expr = not_ilike(col("c1"), null.clone());
+        assert_eq!(simplify(expr), lit_bool_null());
+
+        // expr [NOT] [I]LIKE '%'
+        let expr = like(col("c1"), lit("%"));
+        assert_eq!(simplify(expr), if_not_null(col("c1"), true));
+
+        let expr = not_like(col("c1"), lit("%"));
+        assert_eq!(simplify(expr), if_not_null(col("c1"), false));
+
+        let expr = ilike(col("c1"), lit("%"));
+        assert_eq!(simplify(expr), if_not_null(col("c1"), true));
+
+        let expr = not_ilike(col("c1"), lit("%"));
+        assert_eq!(simplify(expr), if_not_null(col("c1"), false));
+
+        // expr [NOT] [I]LIKE '%%'
```

Comment:
Also, since the code seems to handle more than two `%`, could you add a test for patterns like

- `col LIKE '17%%'      // 17%`
- `col LIKE '%%five%%    // %five%`
- `col LIKE '%%%%five%%%%    // %%five%%`

Reply:
> I thought `expr LIKE '%%'` means the same as `expr = '%'` (match the actual literal character `%` )
> 
> Specifically that `%%` is how to escape the wildcard to not be a wildcard

I think it depends on how we define it. I assume we prefer to follow the Postgres behavior. 🤔 
In Postgres, `%%` doesn't mean matching the actual literal '%'. In Postgres, the escape character is `\`.
```
test=# select '1' like '%%';
 ?column? 
----------
 t
(1 row)

test=# select '%' like '%%';
 ?column? 
----------
 t
(1 row)

test=# select '%' like '\%';
 ?column? 
----------
 t
(1 row)

test=# select '1' like '\%';
 ?column? 
----------
 f
(1 row)
```

I think it is just a redundant wildcard. So, we can simplify it to be only one `%`.


By the way, in DuckDB, there is a similar behavior for double '%%'.
```
D select '1' like '%%';
┌───────────────┐
│ ('1' ~~ '%%') │
│    boolean    │
├───────────────┤
│ true          │
└───────────────┘
D select '%' like '%%';
┌───────────────┐
│ ('%' ~~ '%%') │
│    boolean    │
├───────────────┤
│ true          │
└───────────────┘
````
However, the escape character should be set by `ESCAPE` syntax
```
D select '%' like '\%';
┌───────────────┐
│ ('%' ~~ '\%') │
│    boolean    │
├───────────────┤
│ false         │
└───────────────┘
D select '%' like '\%' escape '\';
┌──────────────────────────────────┐
│ main.like_escape('%', '\%', '\') │
│             boolean              │
├──────────────────────────────────┤
│ true                             │
└──────────────────────────────────┘
```

Reply:
I was definitely confused -- thank you

Reply:
> In Postgres, `%%` doesn't mean matching the actual literal '%'. The escape character is `\`.

thanks @goldmedal for checking this.


in our case, we don't have default escape character at all, i believe.
at least we don't define it on Like object.

i need to check whether we have an _implicit_ default escape character applied somewhere later.

Reply:
The `%%` simplification is wrong. `\` is implicit escape. Let me update.

Reply:
will temporarily rebase on top of https://github.com/apache/datafusion/pull/13288 for more test coverage